pkg/oc/admin/migrate/legacyhpa/hpa.go

```diff
@@ -0,0 +1,209 @@
+		Example: fmt.Sprintf(internalMigrateLegacyHPAExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(options.Complete(name, f, cmd, args))
+			kcmdutil.CheckErr(options.Validate())
+			kcmdutil.CheckErr(options.Run())
+		},
+	}
+	options.Bind(cmd)
+
+	return cmd
+}
+
+func (o *MigrateLegacyHPAOptions) Complete(name string, f *clientcmd.Factory, c *cobra.Command, args []string) error {
+	if len(args) != 0 {
+		return fmt.Errorf("%s takes no positional arguments", name)
+	}
+
+	o.ResourceOptions.SaveFn = o.save
+	if err := o.ResourceOptions.Complete(f, c); err != nil {
+		return err
+	}
+
+	o.finalVersionKinds = make(map[versionKind]versionKind)
+
+	// copy all manual transformations in
+	for initial, final := range defaultMigrations {
+		o.finalVersionKinds[initial] = final
+	}
+
+	kubeClientSet, err := f.ClientSet()
+	if err != nil {
+		return err
+	}
+	o.hpaClient = kubeClientSet.Autoscaling()
+
+	return nil
+}
+
+func (o MigrateLegacyHPAOptions) Validate() error {
+	return o.ResourceOptions.Validate()
+}
+
+func (o MigrateLegacyHPAOptions) Run() error {
+	return o.ResourceOptions.Visitor().Visit(func(info *resource.Info) (migrate.Reporter, error) {
+		return o.checkAndTransform(info.Object)
+	})
+}
+
+func (o *MigrateLegacyHPAOptions) checkAndTransform(hpaRaw runtime.Object) (migrate.Reporter, error) {
+	var hpa *autoscaling.HorizontalPodAutoscaler
+	var wasHPA bool
+	if hpa, wasHPA = hpaRaw.(*autoscaling.HorizontalPodAutoscaler); !wasHPA {
+		return nil, fmt.Errorf("unrecognized object %#v", hpaRaw)
+	}
+
+	currentVersionKind := versionKind{
+		APIVersion: hpa.Spec.ScaleTargetRef.APIVersion,
+		Kind:       hpa.Spec.ScaleTargetRef.Kind,
+	}
+
+	newVersionKind, err := o.latestVersionKind(currentVersionKind)
+	if err != nil {
+		return nil, err
+	}
+
+	if currentVersionKind != newVersionKind {
+		hpa.Spec.ScaleTargetRef.APIVersion = newVersionKind.APIVersion
+		hpa.Spec.ScaleTargetRef.Kind = newVersionKind.Kind
+		return migrate.ReporterBool(true), nil
+	}
+
+	return migrate.ReporterBool(false), nil
+}
+
+func (o *MigrateLegacyHPAOptions) latestVersionKind(current versionKind) (versionKind, error) {
+	if newVersionKind, isKnown := o.finalVersionKinds[current]; isKnown {
+		return newVersionKind, nil
+	}
+
+	return current, nil
+}
+
+// save invokes the API to alter an object. The reporter passed to this method is the same returned by
+// the migration visitor method. It should return an error  if the input type cannot be saved
+// It returns migrate.ErrRecalculate if migration should be re-run on the provided object.
+func (o *MigrateLegacyHPAOptions) save(info *resource.Info, reporter migrate.Reporter) error {
+	var hpa *autoscaling.HorizontalPodAutoscaler
+	var wasHPA bool
+	if hpa, wasHPA = info.Object.(*autoscaling.HorizontalPodAutoscaler); !wasHPA {
+		return fmt.Errorf("unrecognized object %#v", info.Object)
+	}
+
+	updatedHPA, err := o.hpaClient.HorizontalPodAutoscalers(hpa.Namespace).Update(hpa)
```

Comment:
This should just be:

```go
_, err := o.hpaClient.HorizontalPodAutoscalers(hpa.Namespace).Update(hpa)
return migrate.DefaultRetriable(info, err)
```

There is no reason to refresh the info or check the error.

Reply:
ack